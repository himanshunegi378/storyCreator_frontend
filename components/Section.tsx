import React, { RefObject, useRef, useState, useEffect } from 'react'
import useNumberRange from '../hooks/useNumberRange'
import Fragment, { IFragment } from './Fragment'

export interface SectionProps {
  fragments: IFragment[]
}

function Section(props: SectionProps) {
  const [fragmentInViewIndex, setFragmentInViewIndex] = useNumberRange(
    0,
    props.fragments.length
  )

  useEffect(() => {
    const fragmentsCount = props.fragments.length
    if (fragmentInViewIndex > fragmentsCount - 1) {
      setFragmentInViewIndex(fragmentsCount - 1)
    }
    return () => {}
  }, [props.fragments])

  function moveLeft() {
    setFragmentInViewIndex(
      fragmentInViewIndex - 1 >= 0 ? fragmentInViewIndex - 1 : 0
    )
  }

  function moveRight() {
    const fragmentsCount = props.fragments.length
    setFragmentInViewIndex(
      fragmentInViewIndex + 1 <= fragmentsCount - 1
        ? fragmentInViewIndex + 1
        : fragmentsCount - 1
    )
  }

  return (
    <div className="flex w-full justify-between">
      <button
        data-testid="previousButton"
        className="px-2 bg-gray-300"
        onClick={moveLeft}
      >
        {' '}
        {'<'}{' '}
      </button>
      <div data-testid="fragmentView" className="flex-grow">
        {props.fragments[fragmentInViewIndex] ? (
          <Fragment
            text={props.fragments[fragmentInViewIndex].text}
            like={props.fragments[fragmentInViewIndex].like}
          />
        ) : (
          'Add Fragment'
        )}
      </div>
      <button
        data-testid="nextButton"
        className="px-2 bg-gray-300"
        onClick={moveRight}
      >
        {' '}
        {'>'}{' '}
      </button>
    </div>
  )
}

export default Section
