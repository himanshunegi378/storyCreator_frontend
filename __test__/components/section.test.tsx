import React from 'react'
import { cleanup, render } from '@testing-library/react'
import Section, { SectionProps } from '../../components/Section'
import userEvent from '@testing-library/user-event'

afterEach(cleanup)

describe('Section Component', () => {
  it('snapshot with fragments props', () => {
    const sectionProps: SectionProps = {
      fragments: [
        { text: 'himanshu', like: 10 },
        { text: 'himanshu', like: 10 },
        { text: 'himanshu', like: 10 }
      ]
    }
    const { asFragment } = render(<Section {...sectionProps} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('when fragment props empty', () => {
    const sectionProps: SectionProps = {
      fragments: []
    }
    const { getByTestId } = render(<Section {...sectionProps} />)

    expect(getByTestId('fragmentView')).toHaveTextContent('Add Fragment')
  })

  it('should show previous fragment data when previousButton Clicked', () => {
    const sectionProps: SectionProps = {
      fragments: [
        { text: 'Himanshu', like: 10 },
        { text: 'Laxmi', like: 10 },
        { text: 'Lata', like: 10 }
      ]
    }
    const { getByTestId } = render(<Section {...sectionProps} />)
    expect(getByTestId('text')).toHaveTextContent('Himanshu')
    userEvent.click(getByTestId('previousButton'))
    expect(getByTestId('text')).toHaveTextContent('Himanshu')
  })

  it('should show next fragment data when nextButton Clicked', () => {
    const sectionProps: SectionProps = {
      fragments: [
        { text: 'Himanshu', like: 10 },
        { text: 'Laxmi', like: 10 },
        { text: 'Lata', like: 10 }
      ]
    }
    const { getByTestId } = render(<Section {...sectionProps} />)
    expect(getByTestId('text')).toHaveTextContent('Himanshu')
    userEvent.click(getByTestId('nextButton'))
    expect(getByTestId('text')).toHaveTextContent('Laxmi')
  })

  it('Add button click should open modal form to add fragment data', () => {})

  it('modal submision should append a new fragment Data in section props', () => {})
})
