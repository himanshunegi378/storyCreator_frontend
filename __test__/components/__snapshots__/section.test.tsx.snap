// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section Component snapshot with fragments props 1`] = `
<DocumentFragment>
  <div
    class="flex w-full justify-between"
  >
    <button
      class="px-2 bg-gray-300"
      data-testid="previousButton"
    >
       &lt; 
    </button>
    <div
      class="flex-grow"
      data-testid="fragmentView"
    >
      <div
        class="px-4 py-2 bg-blue-50"
      >
        <div
          data-testid="text"
        >
          himanshu
        </div>
        <div
          data-testid="like"
        >
           10 
        </div>
      </div>
    </div>
    <button
      class="px-2 bg-gray-300"
      data-testid="nextButton"
    >
       &gt; 
    </button>
  </div>
</DocumentFragment>
`;
